README
Pg-Pcurse version 0.01
======================

These libraries support pcurse(1), the script that monitors a 
Postgres database cluster, it shows statistics on tables, indexes,
vacuuming, buffers, etc,. 

USAGE

Documentation for usage is described inside the USAGE file, or 
in the pcurse(1) pod .

INSTALLATION

This module depends on Curses and Curses::Widgets . If you
have problems installing them, it is probably you are
missing the header files for libcurses. Debian-bases distributions
will need the libncurses5-dev  package:

$ apt-get install libncurses5-dev

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

This will also install the pcurse(1) script. The first thing you
need to perform is to supply  the correct parameter to pcurse(1)
in order to connect to your Postgres server, something like this:

$ pcurse --host  localhost --user john --passwd apple ,
or
$ pcurse -h localhost --password apple 

---------------------------------------------------------------------
pcurse(1) must connect to the Postgres server in order to do anything.
At startup, pcurse(1) tests if your login is valid before executing any
graphics that might mess up the terminal.  I took great care not to
mess up your screen; in the most unlikely event that the program exits 
abnormally and fails to restore the screen, you can always reset your 
terminal by typing 'reset'. 
----------------------------------------------------------------------

DEPENDENCIES

This module requires these other modules and libraries:

Carp::Assert
Curses
Curses::Widgets
Getopt::Long
DBIx::Abstract
Data::Dumper
File::Slurp
User::pwent




COPYRIGHT AND LICENCE

Copyright (C) 2008 by Ioannis Tambouras

This library is free software; you can redistribute it and/or modify
it under the terms of GPLv3.




Changes
Revision history for Perl extension Pg::Pcurse.

0.23  Sun Aug 29 01:05:01 2011
       - Regular upkeep: changed installer, changed and added tests

0.22  Wed Jul 23 21:48:23 2008
        - Fixed license entry (to perl)

0.20  Thu Jun 26 02:23:59 2008
        - In vacuum mode, Ctrl-L shows configuration settings
        - In vacuum mode, fixed listing of tables so they are ordered
        - In vacuum mode, Ctrl-T shows table statistics
        - In vacuum mode, changed main display to show vac and analyze
        - Old selction values in listboxes are now sticky ( when same params)

0.19  Sat May 31 02:01:56 2008
	- fsm displays names of relations in other db's by its name
          insted of relfileno
	- In buffers mode, Ctrl-L shows fsm relations
	- In buffers mode, Ctrl-T shows fsm and Checkpoint Stats
	- In buffers mode, 'd' shows fsm pages

0.18  Sun May 25 07:47:58 2008
	- Added index definition
	- In statistics mode, added pg_stats for Ctrl-T and Ctrl-L
	- In tables mode, added pg_statio for Ctrl-T 
	- In rules mode, added pg_rewrite for Ctrl-T 
	- Re-arranged buffers mode. Added filters and more info
          in big display; 'd' shows summary; Ctrl-F shows stats and settings.
	- In settings mode, Ctrl-L views fsm rations

0.17  Thu May 22 23:48:15 2008
        - added format info about pgfouine's log_line_prefix
        - Added the missing Defaults.pm file
        - show num of default buffers TODO
        - Ctrl-F in buffers mode, it displays statistics about checkpoints
        - Fixed indexes mode
        - Added 'dict' mode for text search dictionaries

0.15  Thu May 15 20:37:08 2008
        - minor display fixes
        - Ctrl-H displays Help screen
        - Ctrl-F appends the contents of the bottom display to /tmp/pcurse.out
        - Ctrl-L replaces Shift-d , it displays the last 20 updates in table
        - Ctrl-A replaces Shift-a , it analyses  a table or database
        - Ctrl-V replaces Shift-v , it vacuums   a table or database
        - Ctrl-R replaces Shift-r , it reindexes a table or database
        - vacuum, and auto_vacuum dates are properly displayed
        - In settings mode, added 'changed' that selects only the parameters
          set differently than the default postgresql.conf (mine is from Debian)

0.13  Wed May 14 10:28:45 2008
        - fixed use 5.008 in Makefile.PL
	- Mode databases still works, but under development


0.12  Mon May 12 06:52:36 2008
        - schema table is disable for user mode
        - added extra_desc for settings mode
        - In tables mode, 'D' displays the youngest 20 rows of table
        - In rules mode, beautify rule definition
        - In settings mode, settings are now searchable by context
        - In indexes mode, changed or added more fields
        - In procedures mode, 'D' also displays the src

0.11  Wed May  9 06:48:31 2008
        - Display postmaster start time in some modes
        - Display txid_current in some modes
        - Prevent crashes with restricted databases

0.10  Wed May  8 09:39:46 2008
	- In 'database' mode, 'D' displays count of objects
	- In 'settings' mode, 'D' displays fsm rel and pages
	- less clutter in standard 'database' mode
	- bief summary for buffercache (if available)
	- Schema table is disabled when not needed

0.09  Wed May  7 08:55:42 2008
	- minor code refactoring
	- query refactoring

0.08  Man May  5 06:53:22 2008
	- Attemp to read passwd from ~/.pgpass
	- CLI invocation enhancements

0.07  Fri May  2 06:31:03 2008
	- Renamed mode 'overview' to 'databases'
	- Fixed calculation of row sizes 
	- Added more fields to 'd' tables
	- Display updates after analyze & vacuum 

0.06  Man May  2 05:57:37 2008
	- Lost of micro-features, here are some:
	  'd' on schema display shows permissions
	  'D' on big display shows records
	  'A' attempts to analyze a table or database
	  'V' attempts to vacuum a table or database
	  Bucardo, trigger, and views modes
	  Always display what mode we are int
	  Better vacuum submode
	  Prevent terminal blur on failed startup (invalid db credentials) 

0.05  Man Apr 21 03:52:35 2008
	- changed the blue screen of 'tables' to statsof()
	- Fixed module dependacies

0.04  Sun Apr 20 20:39:33 2008
	- Added 'rules' mode
	- Better lexer for sql stataments
	- added ' ' in keyscan for shcema box
	- added 'h' to un-pop the menu display
	- added acl display for schema box
	- The shcema choice is remembered (when possible) accross 
          round of selections
	- Fixed the default mode back to 'tables'

0.03  Sun Apr 20 04:12:49 2008
	- Added 'views' mode
	- procedure arg type are shown in english names

0.02  Sat Apr 19 19:08:32 2008
	- Corrected module dependencies
	- Fixed usage documentation

0.01  Thu Apr 17 23:06:48 2008
	- original version; created by h2xs 1.23 with options
		-AXn Pg::Pcurse



lib/Pg/Pcurse/Widget.pm
# Copyright (C) 2008 Ioannis Tambouras <ioannis@cpan.org>. All rights reserved.
# LICENSE:  GPLv3, eead licensing terms at  http://www.fsf.org .
package Pg::Pcurse::Widget;
use 5.008008;
use Curses;
use Curses::Widgets;
use Carp::Assert;
use Curses::Widgets::Menu;
use Curses::Widgets::Label;
use Curses::Widgets::ButtonSet;
use strict;
use warnings;
use Pg::Pcurse;
our $VERSION = '0.14';


use base 'Exporter';

our @EXPORT = qw( 
	          init_screen       init_mini_root
	          create_root       create_button
		  create_commentbox
	          create_menu
	          create_botton 
	          main_listbox  secondary_listbox  big_listbox
		  form_dbmenu
);

sub miniscan_sec {
	noecho();
        my $mwh = shift;
        my $key = -1;
        while ($key eq -1) {
                $key = $mwh->getch;
                if($key eq "j")  { return KEY_DOWN    };
                if($key eq "k")  { return KEY_UP      };
                if($key eq "h")  { return "\n"        };
                if($key eq ' ')  { return "\n"        };
                if($key eq 'm')  { return KEY_RIGHT   };
                if($key eq 'd')  { got_d($mwh)        };
                if($key eq 'n')  { return KEY_LEFT    };
                if($key eq 'q')  { exit 0             };
        }
        return $key;
}
sub miniscan {
	noecho();
        my $mwh = shift;
        my $key = -1;
        while ($key eq -1) {
                $key = $mwh->getch;
                if($key eq "j")  { return KEY_DOWN    };
                if($key eq "k")  { return KEY_UP      };
                if($key eq "h")  { return "\n"        };
                if($key eq ' ')  { return "\e"        };
                if($key eq 'm')  { return KEY_RIGHT   };
                if($key eq 'n')  { return KEY_LEFT    };
                if($key eq 'q')  { exit 0             };
        }
        return $key;
}
sub miniscan_5c {
	noecho();
        my $mwh = shift;
        my $key = -1;
        while ($key eq -1) {
                $key = $mwh->getch;
                if($key eq "j")  { return KEY_DOWN    };
                if($key eq "k")  { return KEY_UP      };
                if($key eq "h")  { return "\n"        };
                if($key eq ' ')  { return KEY_DOWN    };
                if($key eq 'm')  { return KEY_RIGHT   };
                if($key eq 'n')  { return KEY_LEFT    };
                if($key eq 'q')  { exit 0             };
        }
        return $key;
}


sub _Database_Menu_Choice {
        my $dbs = shift;
        { ITEMORDER => $dbs ,
          map  { my $i=$_; ($_ => sub{ $::db=$i}) } @$dbs,
        }
}


my @MODES = sort qw( Vacuum Stats Procedures Tables Views Users Rules
                     Databases Buffers Indexes Settings Triggers Bucardo
		     Dict
             );
sub form_dbmenu {
        my $dbs = shift;
	assert( ref $dbs, 'ARRAY') if DEBUG;
        my $menus = { MENUORDER => [qw( Databases Mode About ) ],
                      Databases => _Database_Menu_Choice ($dbs),
                      Hide      =>{ ITEMORDER => [ 'System' ],
                                System    => sub { $::hid{system}++} },
                      Mode      =>{ ITEMORDER => [ @MODES ],
                                Vacuum     => sub { $::mode = 'vacuum'    },
                                Stats      => sub { $::mode = 'stats'     },
                                Procedures => sub { $::mode = 'procedures'},
                                Tables     => sub { $::mode = 'tables'    },
                                Views      => sub { $::mode = 'views'     },
                                Users      => sub { $::mode = 'users'     },
                                Databases  => sub { $::mode = 'databases' },
                                Buffers    => sub { $::mode = 'buffers'   },
                                Indexes    => sub { $::mode = 'indexes'   },
                                Dict       => sub { $::mode = 'dict'      },
                                Settings   => sub { $::mode = 'settings'  },
                                Triggers   => sub { $::mode = 'triggers'  },
                                Bucardo    => sub { $::mode = 'bucardo'   },
                                Rules      => sub { $::mode = 'rules'     },
				   },	
                      About      =>{ ITEMORDER => [  
                                            "Version $Pg::Pcurse::VERSION",
                                            'Ioannis Tambouras (C)' 
                                                  ],
                                     },
                    };
        new Curses::Widgets::Menu {
                FOREGROUND  => 'black',
                BACKGROUND  => 'red',
                BORDER      => 1,
                FOCUSSWITCH => "\tl",
	        INPUTFUNC   => \&miniscan  ,
                CURSORPOS   => [qw(Databases)],
                MENUS       => $menus,
         }
}


sub init_screen {
	halfdelay(5);
	curs_set(0);
	leaveok(1);
}

sub create_root {
	my $mwh = new Curses;
	$mwh->erase();
	$mwh->keypad(1);
	$mwh->syncok(1);
	$mwh->attrset(COLOR_PAIR(select_colour(qw(red black))));
	$mwh->box(0,0);
	$mwh->attrset(0);
	$mwh->standout();
	$mwh->standend();
	$mwh;
}


sub create_menu {
	new Curses::Widgets::Menu {
		FOREGROUND  => 'yellow',
		BACKGROUND  => 'green',
		BORDER      => 1,
		CURSORPOS   => [qw(File)],
		MENUS       => { MENUORDER  => [qw(File Help)],
		                 File       => {ITEMORDER=>[qw(Open Save Exit)],
	                                        Open      => sub { 1 },
	                                        Save      => sub { 1 },
	                                        Exit      => sub { exit 0 },
	                       },
		Help    => { ITEMORDER => [qw(Help About)],
		             Help      => sub { 1 },
		             About     => sub { 1 },
		           },
	    },

	  };
}
sub create_botton {
	  new Curses::Widgets::ButtonSet {
		  Y           => 2,
		  X           => 2,
		  FOREGROUND  => 'white',
		  BACKGROUND  => 'black',
		  BORDER      => 0,
		  LABELS      => [ qw( OK CANCEL HELP ) ],
		  LENGTH      => 8,
		  HORIZONTAL  => 1,
	  };
}

sub jscan {
	noecho();
        my $mwh = shift;
        my $key = -1;
        while ($key eq -1) {
                $key = $mwh->getch;
                #if($key eq "\e") {
                        #my $k = $mwh->getch;
                        #if ($k eq 's') { $::mode = 'stats'; return '\t' };
                        #$key = $k; }
                if($key eq 'd' )  { got_h( $mwh )       }
                if($key eq '')  { got_L( $mwh )       }
                if($key eq '')  { got_T( $mwh )       }
                if($key eq '')  { got_H( $mwh )       }
                if($key eq "")  { save2file( $mwh )   }
                if($key eq '')  { analyze             }
                if($key eq '')  { vacuum              }
                if($key eq 'j' )  { return KEY_DOWN     }
                if($key eq 'k' )  { return KEY_UP       }
                if($key eq 'h' )  { return "\n"         }
                if($key eq ' ' )  { return "\n"         }
                if($key eq 'q' )  { exit 0              }
        }
        return $key;
}


sub main_listbox {
	my ($title, $list, $y, $x, $lines) = @_;
	$lines or $lines = @$list;
	assert( ref($list), 'ARRAY') if DEBUG;
	new Curses::Widgets::ListBox {
		  Y           => $y,
		  X           => $x,
		  COLUMNS     => 10,
		  LINES       => $lines,
		  LISTITEMS   => $list,
		  MULTISEL    => 0,
		  VALUE       => 0,
		  INPUTFUNC   => \&miniscan,
		  SELECTEDCOL => 'green',
		  CAPTION     => $title,
		  CAPTIONCOL  => 'yellow',
		  FOCUSSWITCH => "\tl",
		  INPUTFUNC   => \&jscan,
	  };
}
sub secondary_listbox {
	my ($title, $list, $y, $x, $val) = @_;
	#$lines or $lines = @$list;
	assert( ref($list), 'ARRAY') if DEBUG;
	new Curses::Widgets::ListBox {
		  Y           => $y,
		  X           => $x,
		  COLUMNS     => 65,
		  COLUMNS     => 40,
		  LINES       => 7,
		  LISTITEMS   => $list,
		  MULTISEL    => 0,
		  INPUTFUNC   => \&miniscan_sec,
		  FOCUSSWITCH => "\tl",
		  SELECTEDCOL => 'green',
		  CAPTION     => $title,
		  CAPTIONCOL  => 'yellow',
		  CURSORPOS   => $val||0,
		  VALUE       => $val||0,
	  };
}
sub big_listbox {
	my ($title, $list, $y, $x, $val) = @_;
	#$lines or $lines = @$list;
	assert( ref($list), 'ARRAY') if DEBUG;
	new Curses::Widgets::ListBox {
		  Y           => $y,
		  X           => $x,
		  COLUMNS     => 77,
		  LINES       => 12,
		  LISTITEMS   => $list,
		  MULTISEL    => 0,
 		  VALUE       => $val||0,
		  INPUTFUNC   => \&jscan,
		  FOCUSSWITCH => "\tl",
		  CURSORPOS   => $val||0,
		  SELECTEDCOL => 'green',
		  CAPTION     => $title,
		  CAPTIONCOL  => 'yellow',
	  };
}
#####################################################################
sub create_mini_root {
        my $mwh = new Curses @_;
        $mwh->erase();
        $mwh->keypad(1);
        $mwh->syncok(1);
        $mwh->attrset(COLOR_PAIR(select_colour(qw(red black))));
        #$mwh->box(0,0);
        $mwh->attrset(0);
        $mwh->standout();
        $mwh->standend();
        $mwh;
}
our ($sroot, $win_secret);
sub init_mini_root {
	$sroot      = create_mini_root ( 5,40,3,40);
	$win_secret = create_mini_root ( 20,81,4,0);
}

sub got_h {
        my $mwh = shift;
        my $lb_secret  = listbox5 (18,78,0,0, \&retrieve_context)  or return;
        $lb_secret->draw($win_secret,0);
        $lb_secret->execute($win_secret);
}

sub got_d {
        my $mwh = shift;
        my $ll_secret = label_sec( 4,29,0,0) or return;
        $sroot->box(0,0);
        $ll_secret->draw($sroot);
        $ll_secret->execute($sroot);
        sleep 1;
}
sub got_H {
        my $mwh = shift;
        my $ll_secret = label_help( 5,40,0,0) or return;
        $sroot->box(0,0);
        $ll_secret->draw($sroot);
        $ll_secret->execute($sroot);
        sleep 4;
}
sub got_L {
        my $mwh = shift;
        my $lb_secret  = listbox5_white(18,78,0,0,\&capital_context) or return;
        $lb_secret->draw($win_secret,0);
        $lb_secret->execute($win_secret);
}
sub got_T {
        my $mwh = shift;
	eval{
        my $fun = {  tables     => \& table2of   ,
                     stats      => \& stat_of    ,
                     indexes    => \& idx3b      ,
                     vacuum     => \& table3of   ,
                     rules      => \& rewrite_of ,
                     databases  => \& over3      ,
                     buffers    => \& bufcalc    ,
                  }->{$::mode||return};
        my $lb_secret  = listbox5_c2 (18,78,0,0, $fun )  or return;
        $lb_secret->draw($win_secret,0);
        $lb_secret->execute($win_secret);
	} or return;
}
sub display_keyword {
	my $keyword = shift||return;
	my ($y,$x)  = (9,1) ;
	$::mwh->addstr( $y,$x, $keyword);
	$::mwh->refresh;
	sleep 1;
	$::mwh->addstr( $y,$x, ' ' x length$keyword);
	$::mwh->refresh;
}

sub got_A { analyze }
sub got_V { vacuum  }
sub got_R { reindex };

sub listbox5 {
        my ( $lines, $cols, $y,$x, $fun) = @_;
	my $content = $fun->() or return;
        new Curses::Widgets::ListBox {
                  Y           => $x||1,
                  X           => $y||3,
                  COLUMNS     => $cols||25,
                  LISTITEMS   => $content,
                  MULTISEL    => 0,
                  LINES       => $lines||5,
                  INPUTFUNC   => \&miniscan,
                  SELECTEDCOL => 'white',
                  CAPTIONCOL  => 'yellow',
                  FOCUSSWITCH =>  "\tdDl\n",
                  BORDER      => 0,
                  FOREGROUND  => 'white',
                  BACKGROUND  => 'blue',
                  VALUE       =>  0,
          };
}

sub listbox5_c2 {
        my ( $lines, $cols, $y,$x, $fun) = @_;
	my $content = $fun->() or return;
        new Curses::Widgets::ListBox {
                  Y           => $x||1,
                  X           => $y||3,
                  COLUMNS     => $cols||25,
                  LISTITEMS   => $content,
                  MULTISEL    => 0,
                  LINES       => $lines||5,
                  INPUTFUNC   => \&miniscan_5c,
                  SELECTEDCOL => 'black',
                  CAPTIONCOL  => 'yellow',
                  FOCUSSWITCH =>  "\tdDl\n",
                  BORDER      => 0,
                  FOREGROUND  => 'black',
                  BACKGROUND  => 'magenta',
                  VALUE       =>  0,
          };
}
sub listbox5_white {
        my ( $lines, $cols, $y,$x, $fun) = @_;
	my $content = $fun->() or return;
        new Curses::Widgets::ListBox {
                  Y           => $x||1,
                  X           => $y||3,
                  COLUMNS     => $cols||25,
                  LISTITEMS   => $content,
                  MULTISEL    => 0,
                  LINES       => $lines||5,
                  INPUTFUNC   => \&miniscan_5c,
                  SELECTEDCOL => 'black',
                  CAPTIONCOL  => 'yellow',
                  FOCUSSWITCH =>  "\tdDl\n",
                  BORDER      => 0,
                  FOREGROUND  => 'blue',
                  BACKGROUND  => 'white',
                  VALUE       =>  0,
          };
}

sub label_sec {
        my ( $lines, $cols, $y,$x) = @_;
	my $content = retrieve_permit() or return;
        new  Curses::Widgets::Label {
		   COLUMNS     =>  $cols,
		   LINES       =>  $lines,
		   VALUE       =>  "@$content",
		   FOREGROUND  =>  'white',
		   BACKGROUND  =>  'blue',
		   X           =>  $x,
		   Y           =>  $y,
		   ALIGNMENT   => 'C',
        };
}
sub bscan {
        my $mwh = shift;
        my $key = -1;
        while ($key eq -1) {
                $key = $mwh->getch;
                #if($key eq "k")  { return 259};  #ver
                #if($key eq "j")  { return 258};  #ver
                if($key eq "h")  { return 260};  #horz
                if($key eq "k")  { return 261};  #horz
                if($key eq "n")  { return 260};  #horz
                if($key eq "m")  { return 261};  #horz
                if($key eq "l")  { return "\t"};
                if($key eq " ")  { return "\n"};
        }
        return $key;
}

sub create_button {
        my ( $choices, $cols, $x,$y) = @_;
        new Curses::Widgets::ButtonSet   {
           LABELS      => $choices,
           LENGTH      => $cols,
           INPUTFUNC   => \&bscan,
           FOREGROUND  => 'white',
           BACKGROUND  => 'blue',
           BORDER      => 1,
           BORDERCOL   => 'red',
           FOCUSSWITCH => "\t\n",
           HORIZONTAL  => 1,
           PADDING     => 0,
           X           => $x,
           Y           => $y,
        }
}

sub label_help {
        my ( $lines, $cols, $y,$x) = @_;
	my $content = q(
Ctrl-L   Display 20 lastest tuples
Ctrl-T   Statistics
Ctrl-A   Analyze
Ctrl-F   data to File /tmp/pcurse.out 
        );
        new  Curses::Widgets::Label {
		   COLUMNS     =>  $cols,
		   LINES       =>  $lines,
		   VALUE       =>  $content,
		   FOREGROUND  =>  'black',
		   BACKGROUND  =>  'green',
		   X           =>  $x,
		   Y           =>  $y,
		   ALIGNMENT   => 'L',
        };
}
1;

